OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs

```diff
@@ -209,177 +216,71 @@ Edges GetEdges(MPos uv, Func<MPos, bool> isVisible)
-			// tile. We will calculate the vertices for that tile and cache them before drawing it.
-			// Any shroud that is not marked as dirty means our cached tile is still correct - we can just draw the
-			// cached vertices.
-			var visibleUnderShroud = currentShroud.IsExploredTest(visibleRegion);
-			var visibleUnderFog = currentShroud.IsVisibleTest(visibleRegion);
-			foreach (var uv in visibleRegion.MapCoords)
-			{
-				var offset = VertexArrayOffset(uv);
-				if (shroudDirty[uv])
-				{
-					shroudDirty[uv] = false;
-					RenderDirtyTile(uv, offset, visibleUnderShroud, shroudSprites, shroudVertices, shroudPalette, shroudSpriteLayer);
-					RenderDirtyTile(uv, offset, visibleUnderFog, fogSprites, fogVertices, fogPalette, fogSpriteLayer);
-				}
-				else
-				{
-					RenderCachedTile(shroudSpriteLayer[uv], shroudVertices, offset);
-					RenderCachedTile(fogSpriteLayer[uv], fogVertices, offset);
-				}
+				var tileInfo = tileInfos[uv];
+				var shroudSprite = GetSprite(shroudSprites, GetEdges(uv, visibleUnderShroud), tileInfo.Variant);
+				var shroudPos = tileInfo.ScreenPosition;
+				if (shroudSprite != null)
+					shroudPos += shroudSprite.Offset - 0.5f * shroudSprite.Size;
+
+				var fogSprite = GetSprite(fogSprites, GetEdges(uv, visibleUnderFog), tileInfo.Variant);
+				var fogPos = tileInfo.ScreenPosition;
+				if (fogSprite != null)
+					fogPos += fogSprite.Offset - 0.5f * fogSprite.Size;
+
+				shroudLayer.Update(uv, shroudSprite, shroudPos);
+				fogLayer.Update(uv, fogSprite, fogPos);
 			}
-		}
-
-		int VertexArrayOffset(MPos uv)
-		{
-			return 4 * (uv.V * map.MapSize.X + uv.U);
-		}
 
-		void RenderDirtyTile(MPos uv, int offset, Func<MPos, bool> isVisible,
-			Sprite[] sprites, Vertex[] vertices, PaletteReference palette, CellLayer<Sprite> spriteLayer)
-		{
-			var tile = tileInfos[uv];
-			var edges = GetEdges(uv, isVisible);
-			var sprite = CacheTile(uv, offset, edges, tile, sprites, vertices, palette, spriteLayer);
-			RenderCachedTile(sprite, vertices, offset);
-		}
+			cellsDirty.Clear();
+			cellsAndNeighborsDirty.Clear();
 
-		void RenderCachedTile(Sprite sprite, Vertex[] vertices, int offset)
-		{
-			if (sprite != null)
-				Game.Renderer.WorldSpriteRenderer.DrawSprite(sprite, vertices, offset);
-		}
-
-		Sprite CacheTile(MPos uv, int offset, Edges edges, TileInfo tileInfo,
-			Sprite[] sprites, Vertex[] vertices, PaletteReference palette, CellLayer<Sprite> spriteLayer)
-		{
-			var sprite = GetSprite(sprites, edges, tileInfo.Variant);
-			if (sprite != null)
-			{
-				var size = sprite.Size;
-				var location = tileInfo.ScreenPosition - 0.5f * size;
-				OpenRA.Graphics.Util.FastCreateQuad(
-					vertices, location + sprite.FractionalOffset * size,
-					sprite, palette.TextureIndex, offset, size);
-			}
-
-			spriteLayer[uv] = sprite;
-			return sprite;
+			fogLayer.Draw(wr.Viewport);
+			shroudLayer.Draw(wr.Viewport);
```

Comment:
It used to be that fog was rendered after the shroud layer. Might not actually affect things in practise but might be worth keeping that ordering.